modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/PageMemoryMvPartitionStorage.java

```diff
@@ -0,0 +1,515 @@
+            VersionChain versionChain,
+            @Nullable UUID transactionId,
+            @Nullable Timestamp timestamp,
+            Predicate<BinaryRow> keyFilter
+    ) {
+        assert transactionId != null ^ timestamp != null;
+
+        if (transactionId != null) {
+            return findLatestRowVersion(versionChain, transactionId, keyFilter);
+        } else {
+            ByteBufferRow row = findRowVersionByTimestamp(versionChain, timestamp);
+            return keyFilter.test(row) ? row : null;
+        }
+    }
+
+    @Nullable
+    private ByteBufferRow findRowVersionByTimestamp(VersionChain versionChain, Timestamp timestamp) {
+        long nextRowPartitionlessLink = versionChain.headLink();
+
+        while (true) {
+            RowVersion rowVersion = findRowVersion(nextRowPartitionlessLink);
+            Timestamp rowVersionTs = rowVersion.timestamp();
+            if (rowVersionTs != null && rowVersionTs.beforeOrEquals(timestamp)) {
+                return rowVersionToBinaryRow(rowVersion);
+            }
+
+            if (!rowVersion.hasNextLink()) {
+                return null;
+            }
+
+            nextRowPartitionlessLink = rowVersion.nextLink();
+        }
+    }
+
+    @Override
+    public LinkRowId insert(BinaryRow row, UUID txId) throws StorageException {
+        RowVersion rowVersion = insertRowVersion(Objects.requireNonNull(row), RowVersion.NULL_LINK);
+
+        VersionChain versionChain = new VersionChain(partitionId, txId, PartitionlessLinks.removePartitionIdFromLink(rowVersion.link()));
+
+        try {
+            versionChainFreeList.insertDataRow(versionChain);
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Cannot store a version chain", e);
+        }
+
+        try {
+            versionChainTree.putx(versionChain);
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Cannot put a version chain to the tree", e);
+        }
+
+        return new LinkRowId(versionChain.link());
+    }
+
+    private RowVersion insertRowVersion(@Nullable BinaryRow row, long nextPartitionlessLink) {
+        // TODO IGNITE-16913 Add proper way to write row bytes into array without allocations.
+        byte[] rowBytes = row == null ? TOMBSTONE_PAYLOAD : row.bytes();
+
+        RowVersion rowVersion = new RowVersion(partitionId, nextPartitionlessLink, ByteBuffer.wrap(rowBytes));
+
+        insertRowVersion(rowVersion);
+
+        return rowVersion;
+    }
+
+    private void insertRowVersion(RowVersion rowVersion) {
+        try {
+            rowVersionFreeList.insertDataRow(rowVersion);
+        } catch (IgniteInternalCheckedException e) {
+            throw new StorageException("Cannot store a row version", e);
+        }
+    }
+
+    @Override
+    public @Nullable BinaryRow addWrite(RowId rowId, @Nullable BinaryRow row, UUID txId) throws TxIdMismatchException, StorageException {
```

Comment:
As I already mentioned, this method is not effective

Reply:
What do you mean 'not effective'?

Reply:
This is optimized now. BTW, I also optimized commit (only timestamp is written instead of all the version data).